src/persist-client/src/bin/persist_open_loop_benchmark.rs

```diff
@@ -0,0 +1,449 @@
+                            Some(batch) => batch,
+                            None => {
+                                let records_sent = (batches_sent as usize) * args.batch_size;
+                                let finished = format!(
+                                    "Data generator {} finished after {} ms and sent {} records",
+                                    idx, elapsed_millis, records_sent
+                                );
+                                return Ok(finished);
+                            }
+                        };
+                        trace!("data generator {} wrote a batch", idx);
+                        // TODO: clearly specify exact error type. This can only error
+                        // if the receiver has been dropped.
+                        if let Err(_) = tx.send(batch) {
+                            bail!("receiver unexpectedly dropped");
+                        }
+                        batches_sent += 1;
+                    }
+
+                    if elapsed - prev_log > FIVE_SECONDS {
+                        let records_sent = (batches_sent as usize) * args.batch_size;
+                        info!(
+                            "After {} ms data generator {} has sent {} records.",
+                            elapsed_millis, idx, records_sent
+                        );
+                        prev_log = elapsed;
+                    }
+                }
+            },
+        );
+
+        handles.push(data_generator_handle);
+        let b = Arc::clone(&barrier);
+
+        let writer_handle = mz_ore::task::spawn(|| format!("writer-{}", idx), async move {
+            trace!("writer {} waiting for barrier", idx);
+            b.wait().await;
+            info!("starting writer {}", idx);
+
+            // Max observed latency for BenchmarkWriter::write.
+            let mut max_write_latency = Duration::from_millis(0);
+
+            // The last time we emitted progress information to stdout, expressed
+            // as a relative duration from start.
+            let mut prev_log = Duration::from_millis(0);
+            let mut batches_written = 0;
+
+            loop {
+                let batch = match rx.recv().await {
+                    Some(batch) => batch,
+                    None => {
+                        let elapsed = start.elapsed();
+                        let records_written = batches_written * args.batch_size;
+                        let finished = format!("Writer {} finished after {} ms and wrote {} records. Max write latency {} ms.",
+                          idx, elapsed.as_millis(), records_written, max_write_latency.as_millis());
+                        return Ok(finished);
+                    }
+                };
+
+                trace!("writer {} received a batch. writing", idx);
+                let write_start = Instant::now();
+                writer.write(batch).await?;
```

Comment:
its not at all clear to me that the writer should await a response here. Thinking about it overnight I actually think we shouldn't be doing that.

Reply:
Dan and I chatted about this. Conclusion - its desirable to await here because:

1. The writer blocking on writes doesn't affect the open-loop nature of the benchmark. Our end-to-end lag measurements are based off of the reader lag as estimated from the ideal data generation schedule, not the actual rate at which data was generated or the actual rate at which data was written. If instead we had a system where we were measuring lag with respect to how much data was currently written, blocking here would make it closed loop.
2. Blocking makes the semantics clearer and makes it easier to reason about what failed.
3. Measuring write latencies is nice to have.

Conclusion: no changes needed here.